Composition over Inheritance for Hierarchy classes + Default MH Updaters

TeoGiane opened the pull request:
This PR introduce a major code re-factoring for the Hierarchy classes. In order to further maximize the amount of code that can be reused, we changed the code pattern for hierarchies, for which the polymorphic behavior is now achieved via composition.

To be precise, each hierarchy can be seen as the composition of a ``Likelihood``, with obvious meaning (thus managing the state of the hierarchy, I/O methods etc.) and a ``PriorModel``, which manages the hyperparameters and prior of the hierarchy, their update and so on.

This code choice preserves code readability, the front end API for the hierarchies whilst allowing a huge simplification in defining a new Hierarchy, once the ``Likelihood`` and the ``PriorModel`` are already available.

Moreover, this new pattern allows a separate implementation of the update algorithm for the hierarchy. At this stage, thanks to @mberaha contribution to this branch, we have several metropolis updaters that can be used as default update mechanism in a new hierarchy (while allowing possible *ad-hoc* Gibbs-Sampling strategies). This could be a major simplification for end users, which are now not required to implement the strategy for the hierarchy, since they can rely on default updaters.

**Tasks yet to accomplish**
- [x] Finish the porting of ``LinRegUniHierarchy``
- [x] Finish the porting of ``LapNIGHierarchy``
- [x] Finish the porting of ``FAHierarchy``

mberaha commented:
Well this looks like a lot but if you get a closer look @TeoGiane changed only 64 files! 😂😂😂

TeoGiane commented:
> Well this looks like a lot but if you get a closer look @TeoGiane changed only 64 files! joyjoyjoy

Thanks @mberaha, I have not detailed this fact: several changes were needed to use Stan automatic differentiation in Metropolis updaters. To do so, you need to include ``<stan/math/rev.hpp>`` and this should be done BEFORE ``<Eigen/Dense>`` and ``<stan/math/prim.hpp>``. Using directly the ``<stan/math/rev.hpp>``, automatically include the aforementioned libraries.

mberaha reviewed (commented):
I'm working my way through this huge PR. 

I find that before going on, it's better to discuss the use of `prior->get_posterior_hypers()`, specifically: what are these posterior hypers?

mberaha commented on src/hierarchies/likelihoods/laplace_likelihood.cc:
```diff
@@ -0,0 +1,53 @@
+                                         bool add) {
+  if (add) {
+    // sum_abs_diff_curr += std::abs(state.mean - datum(0, 0));
+    cluster_data_values.push_back(datum);
```
I would do it as in the FALikelihood where we only use the index set of the data currently allocated to this cluster and the dataset_ptr

mberaha replied:
(so basically this would be just an empty method)

mberaha commented on src/hierarchies/likelihoods/laplace_likelihood.cc:
```diff
@@ -0,0 +1,53 @@
+
+void LaplaceLikelihood::clear_summary_statistics() {
+  cluster_data_values.clear();
+  // sum_abs_diff_curr = 0;
```
+1

TeoGiane replied:
We now rely on ``dataset_ptr`

mberaha commented on src/hierarchies/likelihoods/laplace_likelihood.h:
```diff
@@ -0,0 +1,56 @@
+#include "base_likelihood.h"
+#include "states/includes.h"
+
+class LaplaceLikelihood
```
docs here

mberaha commented on src/hierarchies/likelihoods/laplace_likelihood.h:
```diff
@@ -0,0 +1,56 @@
+    T mean = unconstrained_params(0);
+    T var = stan::math::positive_constrain(unconstrained_params(1));
+    T out = 0.;
+    for (auto it = cluster_data_values.begin();
+         it != cluster_data_values.end(); ++it) {
+      out += stan::math::double_exponential_lpdf(*it, mean,
+                                                 stan::math::sqrt(var / 2.0));
+    }
```
will need some change since we won't be used `cluster_data_values`

mberaha commented on src/hierarchies/likelihoods/multi_norm_likelihood.h:
```diff
@@ -0,0 +1,46 @@
+#include "base_likelihood.h"
+#include "states/includes.h"
+
+class MultiNormLikelihood
```
docs here

mberaha commented on src/hierarchies/likelihoods/states/multi_ls_state.h:
```diff
@@ -0,0 +1,92 @@
+#include "src/utils/proto_utils.h"
+
+namespace State {
+
```
some docs here

brunoguindani replied:
Also to all other state classes!

mberaha commented on src/hierarchies/likelihoods/states/multi_ls_state.h:
```diff
@@ -0,0 +1,92 @@
+T multi_ls_log_det_jac(
+    Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic> prec_constrained) {
+  T out = 0;
+  stan::math::positive_constrain(stan::math::cov_matrix_free(prec_constrained),
```
I think this is the wrong function, it should be `stan::math::cov_matrix_constrain` or `stan::math::cov_matrix_free`

mberaha replied:
Actually should be `stan::math::cov_matrix_constrain` evaluated on the unconstrained precision, (or -1 * stan::math::cov_matrix_constrain evaluated on the constrained one)

mberaha replied:
It would be good to have some tests!

TeoGiane replied:
> I think this is the wrong function, it should be `stan::math::cov_matrix_constrain` or `stan::math::cov_matrix_free`

This is a HUGE error, thanks!!

> Actually should be stan::math::cov_matrix_constrain evaluated on the unconstrained precision, (or -1 * stan::math::cov_matrix_constrain evaluated on the constrained one)

This will avoid the constrained-unconstrained transformation and save time, I will test it.

mberaha reviewed (commented):
some more stuff!

TeoGiane commented on src/hierarchies/abstract_hierarchy.h:
```diff
@@ -48,16 +51,38 @@
  public:
+  //! Set the likelihood for the current hierarchy. Implemented in the
+  //! BaseHierarchy class
+  virtual void set_likelihood(std::shared_ptr<AbstractLikelihood> like_) = 0;
+
+  //! Set the prior model for the current hierarchy. Implemented in the
+  //! BaseHierarchy class
+  virtual void set_prior(std::shared_ptr<AbstractPriorModel> prior_) = 0;
```
I guess there is no problem in moving directly in ``Basehierarchy``. In this way, we only let the user change the updater via ``set_updater()``, which is still in ``AbstractHierachy``. I am following your suggestion.

TeoGiane commented on src/hierarchies/base_hierarchy.h:
```diff
@@ -22,326 +23,404 @@
+    return like->lpdf_grid(data, covariates);
+  };
+
+  // ADD EXCEPTION HANDLING
```
It is a general doubt rather than a specific TODO, which would eventually require some changes in this "``lpdf`` - like" code patterns. Indeed, using the ``Eigen::RowVectorXd(0)`` as default for the ``covariate`` argument, may produce an unwanted behaviour in case you call, e.g.,  ``Likelihood::lpdf(datum)`` from a dependent hierarchy. You would expect an exception but it would produce an error since it would still call ``lpdf(datum, covariate)``, as ``is_dependent() = true`` and ``covariates = Eigen::RowVectorXd(0)``.

I noticed this issue while testing the univariate linear regression hierarchy, and I set a temporary fix to manage exception correctly in ``src/hierarchies/likelihoods/abstract_likelihood.h:25``. I do not know if this is ideal, but I add those TODO as reminders to share with you.

TeoGiane commented on src/hierarchies/abstract_hierarchy.h:
```diff
@@ -48,16 +51,38 @@
   // EVALUATION FUNCTIONS FOR SINGLE POINTS
   //! Public wrapper for `like_lpdf()` methods
-  double get_like_lpdf(
+  virtual double get_like_lpdf(
```
We override it in ``BaseHierarchy``, but at a closer inspection I guess I can do something better to manage an appropriate method call (also keeping in mind the previous behaviour).

TeoGiane commented on src/hierarchies/priors/base_prior_model.h:
```diff
@@ -0,0 +1,222 @@
+  HyperParams get_hypers() const { return *hypers; }
+
+  //! Returns the struct of the current posterior hyperparameters
+  HyperParams get_posterior_hypers() const { return post_hypers; }
```
> For instance, in the case of a NNIG Hierarchy, they used to be (mupost, lambdapost, apost, bpost) sampled from the conditional distribution given the observations in the cluster.

``post_hypers`` are used to store exactly this stuff. But I think I got your point: not all prior models would be (semi)conjugate. What about moving this stuff inside ``ConjugateUpdater``, which basically computes those posterior hyperparameters?

mberaha replied:
Yes I think this is for the best

mberaha replied:
Yes I think this is for the best

mberaha replied:
I think you can remove this method now

TeoGiane commented on src/hierarchies/fa_hierarchy.h:
```diff
@@ -1,134 +1,52 @@
+    State::FA state;
+    state.mu = hypers.mutilde;
+    state.psi = hypers.beta / (hypers.alpha0 + 1.);
+    state.eta = Eigen::MatrixXd::Zero(hypers.card, hypers.q);
```
This is a leftover from a previous version. I am fixing this. Thanks!

TeoGiane commented on src/hierarchies/likelihoods/base_likelihood.h:
```diff
@@ -0,0 +1,290 @@
+
+namespace internal {
+
+/* SFINAE for cluster_lpdf_from_unconstrained() */
```
Agree. I would also consider moving this stuff either in a separate file or under ``BaseLikelihood`` class (just to have a better code). What do you think?

mberaha replied:
yes I think another file would be great, like `likelihood_helper.h`

TeoGiane replied:
I have now moved this SFINAE stuff in ``likelihood_internal.h``, but I will leave this conversation open as a reminder

TeoGiane commented on src/hierarchies/likelihoods/base_likelihood.h:
```diff
@@ -0,0 +1,290 @@
+    const int id, const Eigen::RowVectorXd &datum,
+    const Eigen::RowVectorXd &covariate) {
+  assert(cluster_data_idx.find(id) == cluster_data_idx.end());
+  card += 1;
+  log_card = std::log(card);
```
Done, thanks!

TeoGiane commented on src/hierarchies/likelihoods/base_likelihood.h:
```diff
@@ -0,0 +1,290 @@
+  static_cast<Derived *>(this)->update_summary_statistics(datum, covariate,
+                                                          false);
+  set_card(card - 1);
+  auto it = cluster_data_idx.find(id);
+  assert(it != cluster_data_idx.end());
```
Here I define ``it`` just because I need it at line 234. Otherwise I need to call ``cluster_data_idx.find(id)`` twice.

mberaha commented on src/hierarchies/nnig_hierarchy.h:
```diff
@@ -1,122 +1,53 @@
+
+  void initialize_state() override {
+    // Get hypers
+    auto hypers = prior->get_hypers();
```
are we using hypers somewhere? like in the BaseHieararchy?

mberaha replied:
are we using hypers somewhere? like in the BaseHieararchy?

mberaha commented on src/hierarchies/priors/abstract_prior_model.h:
```diff
@@ -0,0 +1,96 @@
+
+  //! Sampling from the prior model
+  //! @param use_post_hypers It is a `bool` which decides whether to use prior
+  //! or posterior parameters
```
I would add that this is possible only for conjugate models

TeoGiane replied:
Indeed, we now rely on the version receiving ``hier_hypers`` as input (plus a default behaviour that uses prior hyperparameters). Do we still need this method? I would update the docs and remove this version. Let me know what you think.

mberaha replied:
I don't see the full picture, but if we don't need it, just remove it

mberaha commented on src/hierarchies/priors/abstract_prior_model.h:
```diff
@@ -0,0 +1,96 @@
+
+  //! Sampling from the prior model
+  //! @param use_post_hypers It is a `bool` which decides whether to use prior
+  //! or posterior parameters
```
I would add that this is possible only for conjugate models

mberaha commented on src/hierarchies/priors/fa_prior_model.cc:
```diff
@@ -0,0 +1,178 @@
+}
+
+/*
+// Automatic initialization
```
is the automatic initialization just removed or did you move it to some other place?

TeoGiane replied:
At the moment is commented but available in ``fa_prior_model.cc:190``

mberaha commented on src/hierarchies/priors/fa_prior_model.cc:
```diff
@@ -0,0 +1,178 @@
+}
+
+/*
+// Automatic initialization
```
is the automatic initialization just removed or did you move it to some other place?

mberaha commented on src/hierarchies/updaters/conjugate_updater.h:
```diff
@@ -0,0 +1,57 @@
+#include "src/hierarchies/priors/abstract_prior_model.h"
+
+template <class Likelihood, class PriorModel>
+class ConjugateUpdater : public AbstractUpdater {
```
I would rename this `SemiConjugateUpdater`, since it works also for semi-conjugate hierarchies

mberaha replied:
I would rename this `SemiConjugateUpdater`, since it works also for semi-conjugate hierarchies

mberaha reviewed (commented):
I did not see any major point of concern. I would start polishing the code by removing commented stuff and writing some documentation.

Given the size of the PR, I don't think we need to aim for perfection right now, just to have something that passes all the tests and makes the executables (also the python interface) run fine.

Further refactoring will surely come along the way as we add more priors / hierarchies in the library

TeoGiane commented:
@mberaha With the last three commits I have changed how posterior hypers are managed. Now, we have a ``SemiConjugateUpdater`` class which handles updaters in which the prior and posterior hyperparameters have the same structure. In this way, we manage ``post_hypers`` only in those updaters.

To do so, I had to change some backend interface, now posterior parameters are passed from ``PriorModel``s to ``Updater``s via protocol buffers. This is still not perfect, as I would like to set the use of prior hyperparameters as default behavior of ``PriorModel::sample()``. Before doing so, I would like an opinion.

mberaha commented on src/hierarchies/likelihoods/base_likelihood.h:
```diff
@@ -0,0 +1,290 @@
+  //! the parameter vector can range over (-inf, inf).
+  //! Usually, some kind of transformation is required from the unconstrained
+  //! parameterization to the actual parameterization. This version using
+  //! `stan::math::var` type is required for Stan automatic aifferentiation.
```
typo here

mberaha commented on src/hierarchies/priors/nxig_prior_model.cc:
```diff
@@ -0,0 +1,90 @@
+    bayesmix::AlgorithmState::HierarchyHypers hier_hypers) {
+  auto &rng = bayesmix::Rng::Instance().get();
+  auto params = hier_hypers.nnxig_state();
+
+  double var = stan::math::inv_gamma_rng(params.shape(), params.scale(), rng);
+  double mean = stan::math::normal_rng(params.mean(), sqrt(params.var()), rng);
+
+  bayesmix::AlgorithmState::ClusterState state;
+  state.mutable_uni_ls_state()->set_mean(mean);
+  state.mutable_uni_ls_state()->set_var(var);
+  return std::make_shared<bayesmix::AlgorithmState::ClusterState>(state);
```
I think that a cleaner solution would be

```
State::UniLS out;
out.mean = stan::math::normal_rng(params.mean(), sqrt(params.var()), rng);
out.var = stan::math::inv_gamma_rng(params.shape(), params.scale(), rng);
return out.to_proto()
```

TeoGiane replied:
This is a nice improvement. We have the ``get_as_proto`` method in each ``State`` class, but we do not give any API to the state. Do you think we need to move towards a more structured object (for end users that want to define a new state)?

mberaha replied:
I don't think that is essential right now, I would first merge this PR and then do the polishing

mberaha commented on src/hierarchies/priors/nw_prior_model.cc:
```diff
@@ -175,9 +110,72 @@ void NNWHierarchy::initialize_hypers() {
+  write_prec_to_state(tau_new, &out);
+
+  // Make output state
+  bayesmix::AlgorithmState::ClusterState state;
+  bayesmix::to_proto(out.mean, state.mutable_multi_ls_state()->mutable_mean());
+  bayesmix::to_proto(out.prec, state.mutable_multi_ls_state()->mutable_prec());
+  bayesmix::to_proto(out.prec_chol,
+                     state.mutable_multi_ls_state()->mutable_prec_chol());
+  return std::make_shared<bayesmix::AlgorithmState::ClusterState>(state);
```
I think that `out.to_proto()` should work fine here

mberaha commented:
> This is still not perfect, as I would like to set the use of prior hyperparameters as default behavior of `PriorModel::sample()`. Before doing so, I would like an opinion.

I think that we can make use of `boost::optional` here. Here is a simple example on the usage
```
#include <boost/optional/optional.hpp>

int foo(boost::optional<int> val = boost::optional<int>()) {
  if ( val ) 
    return *val + 2;
  
  else return foo(5);
}

TEST(optional, misc) {
  ASSERT_EQ(foo(1), 3);
  ASSERT_EQ(foo(), 7);
}
```

brunoguindani commented on executables/run_mcmc.cc:
```diff
@@ -168,6 +168,11 @@ int main(int argc, char *argv[]) {
   bayesmix::read_proto_from_file(args.get<std::string>("--hier-args"),
                                  hier->get_mutable_prior());
 
+  // std::cout << "hier->prior: \n"
+  //           << hier->get_mutable_prior()->DebugString() << std::endl;
```
Reminder to remove commented lines

brunoguindani commented on resources/tutorial/algo.asciipb:
```diff
@@ -1,6 +1,6 @@
 ##### GENERIC SETTINGS FOR ALL ALGORITHMS #####
 # Algorithm ID string, e.g. "Neal2"
-algo_id: "Neal3"
+algo_id: "Neal8"
```
We'd make more of a good impression with Neal3 :)

brunoguindani commented on test/lpdf.cc:
```diff
@@ -1,13 +1,13 @@
 #include "src/hierarchies/lin_reg_uni_hierarchy.h"
 #include "src/hierarchies/nnig_hierarchy.h"
-#include "src/hierarchies/nnw_hierarchy.h"
+// #include "src/hierarchies/nnw_hierarchy.h"
```
+1 for commented lines

TeoGiane replied:
Do we have the correct lpdf::nnw test now?

TeoGiane replied:
I checked it out: It requires a lot of work to fix the code to compile test_bayesmix. I would simply remove it, or at least preserve it away from the release.

brunoguindani commented on test/CMakeLists.txt:
```diff
@@ -18,16 +18,19 @@ FetchContent_MakeAvailable(googletest)
   hierarchies.cc
   lpdf.cc
-  priors.cc
+  # priors.cc // OLD, USEREI prior_models.cc
```
+1 for commented lines

brunoguindani commented on src/proto/algorithm_state.proto:
```diff
@@ -42,7 +42,8 @@ message AlgorithmState {
       MultiNormalIGDistribution lin_reg_uni_state = 4;
-      LapNIGState lapnig_state = 6;
+      NxIGDistribution nnxig_state = 5;
+      // LapNIGState lapnig_state = 6;
```
+1 for commented lines

brunoguindani commented on src/proto/hierarchy_id.proto:
```diff
@@ -10,6 +10,7 @@ enum HierarchyId {
     LinRegUni = 3; // Linear Regression (univariate response)
-    LapNIG = 4; // Laplace - Normal Inverse Gamma
-    FA = 5; // Factor Analysers
+    NNxIG = 4; // Normal - Normal x Inverse Gamma
+    LapNIG = 5; // Laplace - Normal Inverse Gamma
+    FA = 6; // Factor Analysers
```
Shouldn't we keep previous numbering as it was?

brunoguindani commented on src/hierarchies/updaters/CMakeLists.txt:
```diff
@@ -0,0 +1,20 @@
+target_sources(bayesmix PUBLIC
+    abstract_updater.h
+    # conjugate_updater.h
```
+1 for commented lines

brunoguindani commented on src/hierarchies/updaters/target_lpdf_unconstrained.cc:
```diff
@@ -0,0 +1,4 @@
+#include "target_lpdf_unconstrained.h"
+
+// target_lpdf_unconstrained::target_lpdf_unconstrained(
+//     AbstractHierarchy *p) : parent(p) {}
```
What's up with this empty file...?

brunoguindani commented on src/utils/distributions.cc:
```diff
@@ -1,10 +1,11 @@
 #include "distributions.h"
 
-#include <Eigen/Dense>
+// #include <Eigen/Dense>
```
+1 for commented lines

brunoguindani commented on test_mh_updater.cpp:
```diff
@@ -0,0 +1,54 @@
+#include <math.h>
```
Do we need this file? especially since it is in the root folder?

TeoGiane replied:
This is @mberaha production, but I guess this is no more needed in the final version.

TeoGiane replied:
I am moving this stuff out from the PR, this is clearly a test. We will be able to resume it at any time.

brunoguindani reviewed (commented):
Part 1 of my review. This was the easy one, as I went through the +1-1 files to clean up so that I can focus on the actually important files. These ones are just reminders for commented lines and little else. I'll start reviewing the remaining 50 files right now!

brunoguindani commented on src/hierarchies/base_hierarchy.h:
```diff
@@ -22,326 +23,415 @@
-
-  //! Pointer to the dataset matrix for the mixture model
-  const Eigen::MatrixXd *dataset_ptr = nullptr;
+  // TEMPORANEO!
+  // const Eigen::MatrixXd *dataset_ptr;
```
+1 for commented lines

brunoguindani commented on src/hierarchies/fa_hierarchy.h:
```diff
@@ -1,134 +1,52 @@
+// #include <Eigen/Dense>
+// #include <memory>
+// #include <vector>
 
-#include "algorithm_state.pb.h"
-#include "base_hierarchy.h"
+// #include "algorithm_state.pb.h"
+// #include "conjugate_hierarchy.h"
 #include "hierarchy_id.pb.h"
-#include "hierarchy_prior.pb.h"
 #include "src/utils/distributions.h"
+// #include "hierarchy_prior.pb.h"
```
+1 for commented lines

brunoguindani commented on src/hierarchies/fa_hierarchy.h:
```diff
@@ -1,134 +1,52 @@
+    State::FA state;
+    state.mu = hypers.mutilde;
+    state.psi = hypers.beta / (hypers.alpha0 + 1.);
+    // state.eta = Eigen::MatrixXd::Zero(hypers.card, hypers.q);
```
+1 for commented lines

brunoguindani commented on examples/gamma_hierarchy/gamma_gamma_hier.h:
```diff
@@ -5,8 +5,8 @@
 #include <src/hierarchies/base_hierarchy.h>
 #include <src/hierarchies/conjugate_hierarchy.h>
```
Please note that the Gamma-Gamma example does not currently compile because of this inclusion

brunoguindani commented on src/hierarchies/likelihoods/abstract_likelihood.h:
```diff
@@ -0,0 +1,187 @@
+  //! Default destructor
+  virtual ~AbstractLikelihood() = default;
+
+  //! Returns an independent, data-less copy of this object. Implemented in
+  //! BaseLikelihood
```
Again, I would remove this "Implemented in BaseLikelihood", also in other points of this file

mberaha replied:
agreed. I get why you put these comments here, but IMO it would be better to have a well written Readme file in the src/hierarchies/likelihoods folder that explains exactly which methods one needs to implement in an actual Likelihood.

TeoGiane replied:
I have removed all these stuff. I will leave this open for the DOCS stuff that @mberaha is suggesting.

brunoguindani commented on src/hierarchies/likelihoods/abstract_likelihood.h:
```diff
@@ -0,0 +1,187 @@
+  //! the parameter vector can range over (-inf, inf).
+  //! Usually, some kind of transformation is required from the unconstrained
+  //! parameterization to the actual parameterization. This version using
+  //! `stan::math::var` type is required for Stan automatic aifferentiation.
```
Typo: "aifferentiation"

brunoguindani commented on src/hierarchies/likelihoods/base_likelihood.h:
```diff
@@ -0,0 +1,290 @@
+  void set_state_from_unconstrained(
+      const Eigen::VectorXd &unconstrained_state) override {
+    internal::set_state_from_unconstrained(state, unconstrained_state, 0);
+    // state.set_from_unconstrained(unconstrained_state);
```
+1 for commented lines

brunoguindani commented on src/hierarchies/likelihoods/base_likelihood.h:
```diff
@@ -0,0 +1,290 @@
+    const Eigen::RowVectorXd &covariate) {
+  assert(cluster_data_idx.find(id) == cluster_data_idx.end());
+  set_card(++card);
+  // card += 1;
+  // log_card = std::log(card);
```
+1 for commented lines

brunoguindani commented on src/hierarchies/likelihoods/base_likelihood.h:
```diff
@@ -0,0 +1,290 @@
+Eigen::VectorXd BaseLikelihood<Derived, State>::lpdf_grid(
+    const Eigen::MatrixXd &data, const Eigen::MatrixXd &covariates) const {
+  Eigen::VectorXd lpdf(data.rows());
+  if (covariates.cols() == 0) {
+    // Pass null value as covariate
+    for (int i = 0; i < data.rows(); i++) {
+      lpdf(i) = static_cast<Derived const *>(this)->lpdf(
+          data.row(i), Eigen::RowVectorXd(0));
+    }
+  } else if (covariates.rows() == 1) {
+    // Use unique covariate
+    for (int i = 0; i < data.rows(); i++) {
+      lpdf(i) = static_cast<Derived const *>(this)->lpdf(data.row(i),
+                                                         covariates.row(0));
+    }
+  } else {
+    // Use different covariates
+    for (int i = 0; i < data.rows(); i++) {
+      lpdf(i) = static_cast<Derived const *>(this)->lpdf(data.row(i),
+                                                         covariates.row(i));
+    }
+  }
```
I have already seen this piece of code multiple times already, would it be possible to recycle it somehow?

mberaha replied:
Good idea, we could do it this way

```
std::function<Eigen::RowVectorXd(int)> cov_getter;
if (covariates.cols() == 0) {
  cov_getter = [](int i){return Eigen::RowVectorXd(0)};
} else if (covariates.rows() == 1) {
  cov_getter = [&](int i){return covariates.row(0)};
} else {
  cov_getter = [&](int i){return covariates.row(i)};
}

for (int i = 0; i < data.rows(); i++) {
       lpdf(i) = static_cast<Derived const *>(this)->lpdf(
              data.row,  cov_getter(i));
 }
```

TeoGiane replied:
Very nice suggestion, @brunoguindani! I will code it as @mberaha says and make all required changes.

brunoguindani commented on src/hierarchies/likelihoods/laplace_likelihood.cc:
```diff
@@ -0,0 +1,54 @@
+      datum(0), state.mean, stan::math::sqrt(state.var / 2.0));
+}
+
+// void LaplaceLikelihood::update_sum_stats(const Eigen::RowVectorXd &datum,
+//                                          bool add) {
+//   if (add) {
+//     sum_abs_diff_curr += std::abs(state.mean - datum(0, 0));
+//     cluster_data_values.push_back(datum);
+//   } else {
+//     sum_abs_diff_curr -= std::abs(state.mean - datum(0, 0));
+//     auto it = std::find(cluster_data_values.begin(),
+//     cluster_data_values.end(),
+//                         datum);
+//     cluster_data_values.erase(it);
+//   }
+// }
```
+1 for commented lines, also below

brunoguindani commented on src/hierarchies/likelihoods/laplace_likelihood.h:
```diff
@@ -0,0 +1,59 @@
+    return;
+  };
+
+  //! Set of values of data points belonging to this cluster
+  // std::list<Eigen::RowVectorXd> cluster_data_values;
+  //! Sum of absolute differences for current params
+  // double sum_abs_diff_curr = 0;
+  //! Sum of absolute differences for proposal params
+  // double sum_abs_diff_prop = 0;
```
+1 for commented lines (are we at like +500 already?)

brunoguindani commented on src/hierarchies/likelihoods/states/multi_ls_state.h:
```diff
@@ -0,0 +1,93 @@
+  return std::make_tuple(mean, prec);
+}
+
+// SEE GitHub for tests
```
? Remove, I guess?

TeoGiane replied:
@mberaha and I were discussing about some tests to improve the computation of ``log_det_jac``. I am leaving this conversation open as a reminder

brunoguindani commented on src/hierarchies/lin_reg_uni_hierarchy.h:
```diff
@@ -1,144 +1,65 @@
 #define BAYESMIX_HIERARCHIES_LIN_REG_UNI_HIERARCHY_H_
 
-#include <google/protobuf/stubs/casts.h>
+// #include <google/protobuf/stubs/casts.h>
```
+1 for commented lines

brunoguindani commented on src/hierarchies/likelihoods/uni_norm_likelihood.cc:
```diff
@@ -0,0 +1,49 @@
+  data_sum_squares = 0;
+}
+
+// double UniNormLikelihood::cluster_lpdf_from_unconstrained(
+//     Eigen::VectorXd unconstrained_params) {
+//   assert(unconstrained_params.size() == 2);
+//   double mean = unconstrained_params(0);
+//   double var = std::exp(unconstrained_params(1));
+//   double out = -(data_sum_squares - 2 * mean * data_sum + card * mean *
+//   mean) /
+//                (2 * var);
+//   out -= card * 0.5 * std::log(stan::math::TWO_PI * var);
+//   return out;
+// }
```
+1 for commented lines

brunoguindani commented on src/hierarchies/nnig_hierarchy.h:
```diff
@@ -1,122 +1,54 @@
 #define BAYESMIX_HIERARCHIES_NNIG_HIERARCHY_H_
 
-#include <google/protobuf/stubs/casts.h>
+// #include <google/protobuf/stubs/casts.h>
```
+1 for commented lines in this file

brunoguindani commented on src/hierarchies/nnw_hierarchy.h:
```diff
@@ -1,168 +1,78 @@
 #define BAYESMIX_HIERARCHIES_NNW_HIERARCHY_H_
 
-#include <google/protobuf/stubs/casts.h>
+// #include <google/protobuf/stubs/casts.h>
```
+1 for commented lines in this file

brunoguindani commented on src/hierarchies/nnw_hierarchy.h:
```diff
@@ -1,168 +1,78 @@
+    unsigned int dim = like->get_dim();
+    double nu_n = params.deg_free() - dim + 1;
+    double coeff = (params.var_scaling() + 1) / (params.var_scaling() * nu_n);
+    // Eigen::MatrixXd scale = bayesmix::to_eigen(params.scale());
```
+1 for commented lines

brunoguindani commented on src/hierarchies/nnxig_hierarchy.h:
```diff
@@ -0,0 +1,43 @@
+#ifndef BAYESMIX_HIERARCHIES_NNXIG_HIERARCHY_H_
+#define BAYESMIX_HIERARCHIES_NNXIG_HIERARCHY_H_
+
+// #include <google/protobuf/stubs/casts.h>
```
+1 for commented lines in this file

brunoguindani reviewed (commented):
Part 2 of my review. Again, mostly pointing out commented lines, but there is some genuine question. Other than that, so far so good. On to the last 30 files!

brunoguindani commented on src/hierarchies/priors/base_prior_model.h:
```diff
@@ -0,0 +1,222 @@
+  //! Container for prior hyperparameters values
+  std::shared_ptr<HyperParams> hypers = std::make_shared<HyperParams>();
+
+  //! Container for posterior hyperparameters values
+  // HyperParams post_hypers;
```
you know by now

brunoguindani commented on src/hierarchies/priors/base_prior_model.h:
```diff
@@ -0,0 +1,222 @@
+  }
+}
+
+// OLD STUFF
```
+1 for commented lines

brunoguindani commented on src/hierarchies/priors/fa_prior_model.cc:
```diff
@@ -0,0 +1,212 @@
+  // Proto2Eigen conversion
+  Eigen::VectorXd mu = bayesmix::to_eigen(state.mu());
+  Eigen::VectorXd psi = bayesmix::to_eigen(state.psi());
+  // Eigen::MatrixXd eta = bayesmix::to_eigen(state.eta());
```
+1 for commented lines in the whole file

TeoGiane replied:
I am letting this conversation open as I need to move a function here still commented.

brunoguindani commented on src/hierarchies/priors/fa_prior_model.h:
```diff
@@ -0,0 +1,47 @@
+#ifndef BAYESMIX_HIERARCHIES_PRIORS_FA_PRIOR_MODEL_H_
+#define BAYESMIX_HIERARCHIES_PRIORS_FA_PRIOR_MODEL_H_
+
+// #include <google/protobuf/stubs/casts.h>
```
+1 for commented lines in the file

brunoguindani commented on src/hierarchies/priors/hyperparams.h:
```diff
@@ -0,0 +1,36 @@
+struct FA {
+  Eigen::VectorXd mutilde, beta;
+  double phi, alpha0;
+  unsigned int /*card,*/ q;
```
Please remove card comment if not needed

brunoguindani commented on src/hierarchies/priors/mnig_prior_model.cc:
```diff
@@ -0,0 +1,112 @@
+  return std::make_shared<bayesmix::AlgorithmState::ClusterState>(state);
+}
+
+// std::shared_ptr<google::protobuf::Message> MNIGPriorModel::sample(
```
+1 for commented lines

brunoguindani commented on src/hierarchies/priors/nig_prior_model.cc:
```diff
@@ -102,22 +70,58 @@ void NNIGHierarchy::initialize_hypers() {
+  return target;
+}
+
+// std::shared_ptr<google::protobuf::Message> NIGPriorModel::sample(
```
+1 for commented lines in the file

brunoguindani commented on src/hierarchies/priors/mnig_prior_model.h:
```diff
@@ -0,0 +1,46 @@
+#ifndef BAYESMIX_HIERARCHIES_PRIORS_MNIG_PRIOR_MODEL_H_
+#define BAYESMIX_HIERARCHIES_PRIORS_MNIG_PRIOR_MODEL_H_
+
+// #include <google/protobuf/stubs/casts.h>
```
+1 for commented lines also below

brunoguindani commented on src/hierarchies/priors/nig_prior_model.h:
```diff
@@ -0,0 +1,59 @@
+#ifndef BAYESMIX_HIERARCHIES_PRIORS_NIG_PRIOR_MODEL_H_
+#define BAYESMIX_HIERARCHIES_PRIORS_NIG_PRIOR_MODEL_H_
+
+// #include <google/protobuf/stubs/casts.h>
```
+1 for commented lines also below

brunoguindani commented on src/hierarchies/priors/nw_prior_model.cc:
```diff
@@ -175,9 +110,72 @@ void NNWHierarchy::initialize_hypers() {
+  return std::make_shared<bayesmix::AlgorithmState::ClusterState>(state);
+};
+
+// std::shared_ptr<google::protobuf::Message> NWPriorModel::sample(
```
uchcudhcieqcjacfcwqtd7hva

brunoguindani commented on src/hierarchies/priors/nw_prior_model.h:
```diff
@@ -0,0 +1,51 @@
+#ifndef BAYESMIX_HIERARCHIES_PRIORS_NW_PRIOR_MODEL_H_
+#define BAYESMIX_HIERARCHIES_PRIORS_NW_PRIOR_MODEL_H_
+
+// #include <google/protobuf/stubs/casts.h>
```
+1 for commented lines in the whole file

brunoguindani commented on src/hierarchies/priors/nxig_prior_model.cc:
```diff
@@ -0,0 +1,90 @@
+  return std::make_shared<bayesmix::AlgorithmState::ClusterState>(state);
+};
+
+// std::shared_ptr<google::protobuf::Message> NxIGPriorModel::sample(
```
senil detnemmoc rof 1+

brunoguindani commented on src/hierarchies/priors/nxig_prior_model.h:
```diff
@@ -0,0 +1,45 @@
+#ifndef BAYESMIX_HIERARCHIES_PRIORS_NXIG_PRIOR_MODEL_H_
+#define BAYESMIX_HIERARCHIES_PRIORS_NXIG_PRIOR_MODEL_H_
+
+// #include <google/protobuf/stubs/casts.h>
```
+1 in the file

brunoguindani commented on src/hierarchies/updaters/abstract_updater.h:
```diff
@@ -0,0 +1,42 @@
+
+#include "src/hierarchies/likelihoods/abstract_likelihood.h"
+#include "src/hierarchies/priors/abstract_prior_model.h"
+// #include "src/hierarchies/updaters/target_lpdf_unconstrained.h"
```
You cannot escape the fate of me reminding you of commented lines

brunoguindani commented on src/hierarchies/updaters/fa_updater.h:
```diff
@@ -0,0 +1,37 @@
+                     const FALikelihood& like);
+  void sample_psi(State::FA& state, const Hyperparams::FA& hypers,
+                  const FALikelihood& like);
+  // void sample_eta(State::FA & state, const Hyperparams::FA & hypers, const
+  // Eigen::MatrixXd * dataset_ptr, const std::set<int> & cluster_data_idx);
+  // void sample_mu(State::FA & state, const Hyperparams::FA & hypers, const
+  // Eigen::VectorXd & data_sum); void sample_lambda(State::FA & state, const
+  // Hyperparams::FA & hypers, const Eigen::MatrixXd * dataset_ptr, const
+  // std::set<int> & cluster_data_idx, size_t dim); void sample_psi(State::FA &
+  // state, const Hyperparams::FA & hypers, const Eigen::MatrixXd *
+  // dataset_ptr, const std::set<int> & cluster_data_idx, size_t dim);
```
Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.

brunoguindani commented on src/hierarchies/updaters/mala_updater.h:
```diff
@@ -0,0 +1,58 @@
+    }
+    Eigen::VectorXd grad;
+    double tmp;
+    stan::math::gradient(target_lpdf, curr_state, tmp, grad);
```
Cool!!! 🤩

brunoguindani commented on src/hierarchies/updaters/metropolis_updater.h:
```diff
@@ -0,0 +1,32 @@
+#include "abstract_updater.h"
+#include "target_lpdf_unconstrained.h"
+
+template <class DerivedUpdater>
+class MetropolisUpdater : public AbstractUpdater {
```
I would add some doc in both this class and MALAUpdater

brunoguindani commented on src/hierarchies/updaters/mnig_updater.h:
```diff
@@ -0,0 +1,23 @@
+  ProtoHypers compute_posterior_hypers(AbstractLikelihood &like,
+                                       AbstractPriorModel &prior) override;
+
+  // void compute_posterior_hypers(AbstractLikelihood& like,
+  //                               AbstractPriorModel& prior) override;
```
I've run out of clever jokes

brunoguindani commented on src/hierarchies/updaters/mnig_updater.cc:
```diff
@@ -0,0 +1,83 @@
+
+  var_scaling = covar_sum_squares + hypers.var_scaling;
+  auto llt = var_scaling.llt();
+  // var_scaling_inv = llt.solve(Eigen::MatrixXd::Identity(dim, dim));
```
+1 for commented lines also below

brunoguindani commented on src/hierarchies/updaters/nnig_updater.h:
```diff
@@ -0,0 +1,23 @@
+  ProtoHypers compute_posterior_hypers(AbstractLikelihood &like,
+                                       AbstractPriorModel &prior) override;
+
+  // void compute_posterior_hypers(AbstractLikelihood& like,
+  //                               AbstractPriorModel& prior) override;
```
"Why are we still here? Just to suffer?"

TeoGiane replied:
Guess I am a coding version of a hoarder LOL

brunoguindani commented on src/hierarchies/updaters/nnig_updater.cc:
```diff
@@ -0,0 +1,80 @@
+  out.mutable_nnig_state()->set_var_scaling(var_scaling);
+  out.mutable_nnig_state()->set_shape(shape);
+  out.mutable_nnig_state()->set_scale(scale);
+  // priorcast.set_posterior_hypers(post_params);
```
C O M M E N T S

brunoguindani commented on src/hierarchies/updaters/nnw_updater.h:
```diff
@@ -0,0 +1,23 @@
+  ProtoHypers compute_posterior_hypers(AbstractLikelihood &like,
+                                       AbstractPriorModel &prior) override;
+
+  // void compute_posterior_hypers(AbstractLikelihood& like,
+  //                               AbstractPriorModel& prior) override;
```
!

brunoguindani commented on src/hierarchies/updaters/nnw_updater.cc:
```diff
@@ -0,0 +1,88 @@
+              (mubar - hypers.mean) * (mubar - hypers.mean).transpose();
+  scale_inv = tau_temp + hypers.scale_inv;
+  scale = stan::math::inverse_spd(scale_inv);
+  // scale_chol = Eigen::LLT<Eigen::MatrixXd>(scale).matrixU();
```
also below

brunoguindani commented on src/hierarchies/updaters/nnxig_updater.h:
```diff
@@ -0,0 +1,21 @@
+  ProtoHypers compute_posterior_hypers(AbstractLikelihood &like,
+                                       AbstractPriorModel &prior) override;
+
+  // void compute_posterior_hypers(AbstractLikelihood& like,
+  //                               AbstractPriorModel& prior) override;
```
🅱️

brunoguindani commented on src/hierarchies/updaters/nnxig_updater.cc:
```diff
@@ -0,0 +1,72 @@
+  return out;
+}
+
+// void NNxIGUpdater::compute_posterior_hypers(AbstractLikelihood& like,
```
⌛️ 🎪 📻 🆒 🛄 🌺 🚢 ♻️ 🍉 💮 🔸 🐙 🍼 👎 📹 🍋 👄 😭 🎏 ♥️

brunoguindani reviewed (commented):
Part 3! Not much else to say! Actually, I'll post my high-level thoughts in a bit, if I have any.

brunoguindani commented:
Part 4 out 4 of four, that is, the last one! I really like the new high-level structure. I have just a small question: why putting the states in a subfolder of `hierarchies/likelihoods` rather than `hierarchies`? Also, after dealing with all commented lines, please mark conversations as resolved, or even delete the message, so that this conversation is cleaner.

Now, if you will, I'm gonna go and sleep for 2 days.

TeoGiane commented:
With these last commits I gave a general clean to the whole code (thanks to @brunoguindani suggestions). Some stuff still needs to be done, I will go as quickly as possible and keep you up to date as I add more commits.

mberaha commented on src/hierarchies/base_hierarchy.h:
```diff
@@ -22,326 +22,384 @@
+      const Eigen::MatrixXd &covariates /*= Eigen::MatrixXd(0, 0)*/)
+      const override {
+    Eigen::VectorXd lpdf(data.rows());
+    if (covariates.cols() == 0) {
+      // Pass null value as covariate
+      for (int i = 0; i < data.rows(); i++) {
+        lpdf(i) = static_cast<Derived const *>(this)->conditional_pred_lpdf(
+            data.row(i), Eigen::RowVectorXd(0));
+      }
+    } else if (covariates.rows() == 1) {
+      // Use unique covariate
+      for (int i = 0; i < data.rows(); i++) {
+        lpdf(i) = static_cast<Derived const *>(this)->conditional_pred_lpdf(
+            data.row(i), covariates.row(0));
+      }
+    } else {
+      // Use different covariates
+      for (int i = 0; i < data.rows(); i++) {
+        lpdf(i) = static_cast<Derived const *>(this)->conditional_pred_lpdf(
+            data.row(i), covariates.row(i));
```
I think the same trick with the `cov_getter` lambda function can be used here

mberaha commented on src/hierarchies/base_hierarchy.h:
```diff
@@ -22,326 +22,384 @@
+      const Eigen::MatrixXd &covariates = Eigen::MatrixXd(0, 0)) override {
+    like->clear_data();
+    like->clear_summary_statistics();
+    if (covariates.cols() == 0) {
+      // Pass null value as covariate
+      for (int i = 0; i < data.rows(); i++) {
+        static_cast<Derived *>(this)->add_datum(i, data.row(i), false,
+                                                Eigen::RowVectorXd(0));
+      }
+    } else if (covariates.rows() == 1) {
+      // Use unique covariate
+      for (int i = 0; i < data.rows(); i++) {
+        static_cast<Derived *>(this)->add_datum(i, data.row(i), false,
+                                                covariates.row(0));
+      }
+    } else {
+      // Use different covariates
+      for (int i = 0; i < data.rows(); i++) {
+        static_cast<Derived *>(this)->add_datum(i, data.row(i), false,
+                                                covariates.row(i));
+      }
+    }
```
and here

mberaha commented on src/hierarchies/updaters/semi_conjugate_updater.h:
```diff
@@ -0,0 +1,65 @@
+  if (likecast.get_card() == 0) {
+    likecast.set_state_from_proto(*priorcast.sample(), !set_card);
+  } else {
+    auto post_params = compute_posterior_hypers(likecast, priorcast);
+    likecast.set_state_from_proto(*priorcast.sample(post_params), !set_card);
+    if (update_params) save_posterior_hypers(post_params);
```
Since this is a template class, I think we could avoid having the back-and-forth conversion to protos.

In fact, we could have that the `Updater::compute_posterior_hypers` return an object in the prior/hyperparams namespace. 

Then every PriorModel implements the method

```
SomeState sample(const SomeHypers& params);
```

In the BasePriorModel we provide a unique function with signature
```
std::shared_ptr<google::protobuf::Message> sample(
       ProtoHypersPtr hier_hypers = nullptr) {
   auto params static_cast<Derived*>(this)->hypers_from_proto(hier_hypers);
   auto out = params static_cast<Derived*>(this)->draw(params);
   return out.to_proto();
}
```
the `hypers_from_proto` is basically the same method as `set_hypers_from_proto` but it returns the actual object (or a pointer or whatever is easier).

What do you think?

mberaha replied:
I don't think this is needed right now, but it would make code easier to read. Maybe once we merge the PR we can open an issue and do this when we get time

TeoGiane replied:
This is a huge +1 for me!

mberaha reviewed (commented):
I think we're almost ready to go.

Apart from the very few dangling comments, the last thing missing is the documentation. If it's ok with @brunoguindani I would merge this PR and then split the documentation is smaller PRs that can be done independently.

TeoGiane commented:
**Tasks accomplished from the last review**:
- [x] I have introduced the ``to_proto()`` method in some ``State``s class. This makes code more readable, as suggested by @mberaha
- [x] I have introduced a new callable class, named ``covariates_getter``, which is used to avoid code repetitions in ``*_grid`` methods in ``Hierarchy`` and ``Likelihood`` classes, as suggested by @brunoguindani.
- [x] Removed ``test_mh_updater`` from the parent folder, as it is no more necessary (we can think about a possible different location, if we want to preserve it).

**Things still left behind**:
- [ ] Clean the code I commented during these last commits (which is still to let you see the changes).

**Things I would split in future PRs**
- Documentation
- Improve the ``State``s classes and give them a more rigid structure (this is my idea);
- Avoid back and forth proto conversions as suggested by @mberaha in a still unresolved conversation;

brunoguindani commented:
Does the code run fine? If so, I'm fine with merging this, then polishing with other PRs. Before that, please make sure the Gamma-Gamma example works -- I see the conversation is still unresolved

mberaha commented on src/utils/covariates_getter.h:
```diff
@@ -0,0 +1,27 @@
+#define BAYESMIX_SRC_UTILS_COVARIATES_GETTER_H
+
+#include <Eigen/Dense>
+// #include "src/hierarchies/likelihoods/abstract_likelihood.h"
+// #include "src/hierarchies/priors/abstract_prior_model.h"
```
also this class should have a little documentation

mberaha reviewed (approved):
It's good for me! Huge effort, thanks @TeoGiane 

I'm never ever ever ever reviewing a PR this big in my lifetime

brunoguindani reviewed (approved):
> Upon your approval, this merge will be pushed to `hier-comp-plus-updaters` branch inside `bayesmix-dev`, in order to manage future developments before moving definitively to `master`.

Sure, go ahead :)
